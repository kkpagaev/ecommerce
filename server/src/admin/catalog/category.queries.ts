/** Types generated for queries found in "src/admin/catalog/category.sql" */
import { PreparedQuery } from "@pgtyped/runtime";

/** 'ListCategories' parameters type */
export type IListCategoriesParams = void;

/** 'ListCategories' return type */
export interface IListCategoriesResult {
  description: string | null;
  id: number;
  name: string;
  slug: string;
}

/** 'ListCategories' query type */
export interface IListCategoriesQuery {
  params: IListCategoriesParams;
  result: IListCategoriesResult;
}

const listCategoriesIR: any = {
  usedParamSet: {},
  params: [],
  statement: "SELECT id, name, slug, description FROM categories\nORDER BY id",
};

/**
 * Query generated from SQL:
 * ```
 * SELECT id, name, slug, description FROM categories
 * ORDER BY id
 * ```
 */
export const listCategories = new PreparedQuery<
  IListCategoriesParams,
  IListCategoriesResult
>(listCategoriesIR);

/** 'FindCategoryById' parameters type */
export interface IFindCategoryByIdParams {
  id?: number | null | void;
}

/** 'FindCategoryById' return type */
export interface IFindCategoryByIdResult {
  description: string | null;
  id: number;
  name: string;
  slug: string;
}

/** 'FindCategoryById' query type */
export interface IFindCategoryByIdQuery {
  params: IFindCategoryByIdParams;
  result: IFindCategoryByIdResult;
}

const findCategoryByIdIR: any = {
  usedParamSet: { id: true },
  params: [
    {
      name: "id",
      required: false,
      transform: { type: "scalar" },
      locs: [{ a: 62, b: 64 }],
    },
  ],
  statement:
    "SELECT id, name, slug, description FROM categories\nWHERE id = :id",
};

/**
 * Query generated from SQL:
 * ```
 * SELECT id, name, slug, description FROM categories
 * WHERE id = :id
 * ```
 */
export const findCategoryById = new PreparedQuery<
  IFindCategoryByIdParams,
  IFindCategoryByIdResult
>(findCategoryByIdIR);

/** 'CreateCategory' parameters type */
export interface ICreateCategoryParams {
  description?: string | null | void;
  name?: string | null | void;
  slug?: string | null | void;
}

/** 'CreateCategory' return type */
export interface ICreateCategoryResult {
  id: number;
}

/** 'CreateCategory' query type */
export interface ICreateCategoryQuery {
  params: ICreateCategoryParams;
  result: ICreateCategoryResult;
}

const createCategoryIR: any = {
  usedParamSet: { name: true, slug: true, description: true },
  params: [
    {
      name: "name",
      required: false,
      transform: { type: "scalar" },
      locs: [{ a: 61, b: 65 }],
    },
    {
      name: "slug",
      required: false,
      transform: { type: "scalar" },
      locs: [{ a: 68, b: 72 }],
    },
    {
      name: "description",
      required: false,
      transform: { type: "scalar" },
      locs: [{ a: 75, b: 86 }],
    },
  ],
  statement:
    "INSERT INTO categories\n  (name, slug, description)\nVALUES\n  (:name, :slug, :description)\nRETURNING id",
};

/**
 * Query generated from SQL:
 * ```
 * INSERT INTO categories
 *   (name, slug, description)
 * VALUES
 *   (:name, :slug, :description)
 * RETURNING id
 * ```
 */
export const createCategory = new PreparedQuery<
  ICreateCategoryParams,
  ICreateCategoryResult
>(createCategoryIR);
