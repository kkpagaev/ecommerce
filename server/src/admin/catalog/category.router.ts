import { z } from "zod";
import { publicProcedure } from "../../trpc";
import { FastifyInstance } from "fastify";

type User = {
  id: number;
  name: string;
  bio?: string;
};

const users: Record<string, User> = {};

export default async function (f: FastifyInstance) {
  return f.trpc({
    getUserById: publicProcedure
      .input(z.number()) // input type is string
      .query((opts) => {
        return users[opts.input]; // input type is string
      }),
    createUser: publicProcedure
      .input(
        z.object({
          name: z.string().min(3),
          bio: z.string().max(142).optional(),
        }),
      )
      .mutation((opts) => {
        const id = +Date.now().toString();
        const user: User = { id, ...opts.input };

        users[user.id] = user;

        return user;
      }),
  });
}
