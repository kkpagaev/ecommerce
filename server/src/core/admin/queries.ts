import { sql } from "@pgtyped/runtime";
import { IAdminListCountQueryQuery, IAdminListQueryQuery, IAdminDeleteQueryQuery, IAdminFindOneQueryQuery, IAdminUpdateQueryQuery, IAdminCreateQueryQuery } from "./queries.types";

// | Column     | Type                        | Modifiers                                           |
// |------------|-----------------------------|-----------------------------------------------------|
// | id         | integer                     |  not null default nextval('admin_id_seq'::regclass) |
// | name       | character varying(255)      |                                                     |
// | surname    | character varying(255)      |                                                     |
// | password   | character varying(255)      |  not null                                           |
// | email      | character varying(255)      |  not null                                           |
// | created_at | timestamp without time zone |  not null default CURRENT_TIMESTAMP                 |

export const adminCreateQuery = sql<IAdminCreateQueryQuery>`
  INSERT INTO admin (email, password, name, surname)
  VALUES (
    $email!,
    $password!,
    $name,
    $surname
  )
  RETURNING id
`;

export const adminUpdateQuery = sql<IAdminUpdateQueryQuery>`
  UPDATE admin
  SET
    name = COALESCE($name, name),
    surname = COALESCE($surname, surname),
    email = COALESCE($email, email),
    password = COALESCE($password, password)
  WHERE
    id = $id!
  RETURNING id
`;

export const adminFindOneQuery = sql<IAdminFindOneQueryQuery>`
  SELECT * FROM admin
  WHERE id = COALESCE($id, id)
  AND email = COALESCE($email, email)
`;

export const adminDeleteQuery = sql<IAdminDeleteQueryQuery>`
  DELETE FROM admin
  WHERE id = $id!
`;

export const adminListQuery = sql<IAdminListQueryQuery>`
  SELECT * FROM admin
  LIMIT COALESCE($limit, 10)
  OFFSET (COALESCE($page, 1) - 1) * COALESCE($limit, 10);
`;

export const adminListCountQuery = sql<IAdminListCountQueryQuery>`
  SELECT COUNT(*) FROM admin;
`;

export const adminQueries = {
  admin: {
    listCount: adminListCountQuery,
    list: adminListQuery,
    delete: adminDeleteQuery,
    findOne: adminFindOneQuery,
    update: adminUpdateQuery,
    create: adminCreateQuery,
  },
};
