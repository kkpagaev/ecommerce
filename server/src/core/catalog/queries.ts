import { sql } from "@pgtyped/runtime";
import { IAttributeFindOneQueryQuery, IAttributeValueDeleteManyQueryQuery, IAttributeValueIdListQueryQuery, IProductUpdateQueryQuery, IProductDeleteQueryQuery, IProductAttributeValueListQueryQuery, IProductFindOneQueryQuery, IProductAttributeVAlueInsertQueryQuery, IProductAttributeValueDeleteQueryQuery, IProductCreateQueryQuery, IPriceUpsertQueryQuery, IProductListCountQueryQuery, IProductListQueryQuery, IProductFindByIdQueryQuery, IAttributeValueDeleteQueryQuery, IAttributeListCountQueryQuery, IAttributeListQueryQuery, IAttributeDeleteQueryQuery, IAttributeUpdateQueryQuery, IAttributeCreateQueryQuery, IAttributeValueCreateQueryQuery, IAttributeValueListQueryQuery, ICategoryCreateQueryQuery, ICategoryFindByIdQueryQuery, ICategoryUpdateQueryQuery, ICategoryListCountQueryQuery, ICategoryListQueryQuery, IAttributeValueUpdateQueryQuery, Json } from "./queries.types";
import { PoolClient } from "pg";

export const attributeFindByIdQuery = sql`
 SELECT id, name, description 
 FROM attributes
 WHERE id = $id!
`;

export const attributeFindOneQuery = sql<IAttributeFindOneQueryQuery>`
 SELECT id, name, description
 FROM attributes
 WHERE id = COALESCE($id, id)
 LIMIT 1;
`;

export const attributeListCountQuery = sql<IAttributeListCountQueryQuery>`
 SELECT COUNT(*) FROM attributes;
`;

export const attributeListQuery = sql<IAttributeListQueryQuery>`
 SELECT id, name, description
 FROM attributes
 ORDER BY id
 LIMIT COALESCE($limit, 10)
 OFFSET (COALESCE($page, 1) - 1) * COALESCE($limit, 10);
`;

export const attributeCreateQuery = sql<IAttributeCreateQueryQuery>`
  INSERT INTO attributes
    (name, description)
  VALUES
    $$values(name!, description)
  RETURNING id;
`;

export const attributeUpdateQuery = sql<IAttributeUpdateQueryQuery>`
  UPDATE attributes
  SET
    name = COALESCE($name, name),
    description = COALESCE($description, description)
  WHERE
    id = $id!;
`;

export const attributeDeleteQuery = sql<IAttributeDeleteQueryQuery>`
  DELETE FROM attributes
  WHERE id = $id!;
`;

export const attributeValueListQuery = sql<IAttributeValueListQueryQuery>`
  SELECT id, value, attribute_id
  FROM attribute_values
  WHERE attribute_id = $attribute_id
  ORDER BY id;
`;

export const attributeValueIdListQuery = sql<IAttributeValueIdListQueryQuery>`
  SELECT id FROM attribute_values
  WHERE attribute_id = $attributeId!;
`;

export const attributeValueCreateQuery = sql<IAttributeValueCreateQueryQuery>`
  INSERT INTO attribute_values
    (attribute_id, value)
  VALUES
    $$values(attributeId, value)
  RETURNING id;
`;

export const attributeValueUpdateQuery = sql<IAttributeValueUpdateQueryQuery>`
  UPDATE attribute_values
  SET
    value = COALESCE($value, value)
  WHERE
    id = $id!;
`;

export const attributeValueDeleteQuery = sql<IAttributeValueDeleteQueryQuery>`
  DELETE FROM attribute_values
  WHERE id = $id!;
`;

export const attributeValueDeleteManyQuery = sql<IAttributeValueDeleteManyQueryQuery>`
  DELETE FROM attribute_values
  WHERE id IN $$ids;
`;

export interface IAttributeValueUpsertQueryParams {
  values: readonly ({
    id: number | null | void;
    attributeId: number;
    value: Json;
  })[];
}

// basicaly when inserting null as id, it uses autoincrement
export async function attributeValueUpsertQuery(
  { values }: IAttributeValueUpsertQueryParams,
  client: PoolClient,
): Promise<any> {
  return await client.query(
    `
  INSERT INTO attribute_values
  (id, attribute_id, value)
  VALUES ${
    // ???
    values.map(
      (_, i) => `(
      COALESCE(
        $${i * 3 + 1},
        nextval('attributes_id_seq'::regclass)
      ),
      $${i * 3 + 2}, 
      $${i * 3 + 3})`
    )
  }
  ON CONFLICT (id)
  DO UPDATE SET value = EXCLUDED.value;
    `,
    values.map((i) => [i.id, i.attributeId, i.value]).flat()
  );
}

export const categoryListCountQuery = sql<ICategoryListCountQueryQuery>`
  SELECT COUNT(*) FROM categories;
`;

export const categoryListQuery = sql<ICategoryListQueryQuery>`
  SELECT id, name, slug, description
  FROM categories
  ORDER BY id
  LIMIT COALESCE($limit, 10)
  OFFSET (COALESCE($page, 1) - 1) * COALESCE($limit, 10);
`;

export const categoryFindByIdQuery = sql<ICategoryFindByIdQueryQuery>`
  SELECT id, name, slug, description FROM categories
  WHERE id = $id;
`;

// export const categoryFindQuery = sql<ICategoryFindQueryQuery>`
//   SELECT id, name, slug, description FROM categories
//   WHERE id = COALESCE($id, id)
//   AND name->>$locale! = COALESCE($name, name->>$locale!)
//   AND slug = COALESCE($slug, slug)
// `;

export const categoryCreateQuery = sql<ICategoryCreateQueryQuery>`
  INSERT INTO categories
    (name, slug, description)
  VALUES
    ($name!, $slug!, $description)
  RETURNING id;
`;

export const categoryUpdateQuery = sql<ICategoryUpdateQueryQuery>`
  UPDATE categories
  SET
    name = COALESCE($name, name),
    slug = COALESCE($slug, slug),
    description = COALESCE($description, description)
  WHERE
    id = $id!;
`;

// export const optionCreateQuery = sql<IOptionCreateQueryQuery>`
//   INSERT INTO options
//     (name)
//   VALUES
//     ($name!)
//   RETURNING id;
// `;

// export const optionUpdateQuery = sql<IOptionUpdateQueryQuery>`
//   UPDATE options
//   SET
//     name = COALESCE($name, name)
//   WHERE
//     id = $id!;
// `;

// export const optionDeleteQuery = sql<IOptionDeleteQueryQuery>`
//   DELETE FROM options
//   WHERE
//     id = $id;
// `;

// export const optionFindByIdQuery = sql<IOptionFindByIdQueryQuery>`
//   SELECT id, name
//   FROM options
//   WHERE id = $id;
// `;

export const productListQuery = sql<IProductListQueryQuery>`
  SELECT * FROM products
  ORDER BY id
  LIMIT COALESCE($limit, 10)
  OFFSET (COALESCE($page, 1) - 1) * COALESCE($limit, 10);
`;

export const productFindOneQuery = sql<IProductFindOneQueryQuery>`
  SELECT * FROM products
  WHERE id = COALESCE($id, id)
  LIMIT 1;
`;

export const productDeleteQuery = sql<IProductDeleteQueryQuery>`
  DELETE FROM products
  WHERE id = $id!
  RETURNING id;
`;

export const productCreateQuery = sql<IProductCreateQueryQuery>`
  INSERT INTO products
  (name, description, slug, category_id)
  VALUES
  ($name!, $description, $slug!, $categoryId!)
  RETURNING id;
`;

export const productUpdateQuery = sql<IProductUpdateQueryQuery>`
  UPDATE products
  SET
    name = COALESCE($name, name),
    description = COALESCE($description, description),
    slug = COALESCE($slug, slug),
    category_id = COALESCE($categoryId, category_id)
  WHERE
    id = $id!;
`;

export const productAttributeValueListQuery = sql<IProductAttributeValueListQueryQuery>`
  SELECT v.*,
     a.NAME AS attribute_name
  FROM   attribute_values v
     JOIN product_attributes pa
       ON pa.attribute_value_id = v.id
     JOIN attributes a
       ON a.id = v.attribute_id
  WHERE  pa.product_id = $productId
`;

export const productAttributeValueDeleteQuery = sql<IProductAttributeValueDeleteQueryQuery>`
  DELETE FROM product_attributes WHERE product_id = $product_id;
`;

export const productAttributeVAlueInsertQuery = sql<IProductAttributeVAlueInsertQueryQuery>`
  INSERT INTO product_attributes(product_id, attribute_value_id) 
  VALUES
  $$values(productId!, attributeValueId!);
`;

export const productListCountQuery = sql<IProductListCountQueryQuery>`
  SELECT COUNT(*) FROM products
`;

export const productFindByIdQuery = sql<IProductFindByIdQueryQuery>`
  SELECT * FROM products
  WHERE id = $id;
`;

export const priceUpsertQuery = sql<IPriceUpsertQueryQuery>`
  INSERT INTO prices (product_id, price, type)
  VALUES $$values(product_id!, price!, type)
  ON CONFLICT (product_id, type)
  DO UPDATE SET price = EXCLUDED.price;
`;

export const catalogQueries = {
  category: {
    create: categoryCreateQuery,
    update: categoryUpdateQuery,
    list: categoryListQuery,
    listCount: categoryListCountQuery,
    findById: categoryFindByIdQuery,
  },
  attribute: {
    findOne: attributeFindOneQuery,
    list: attributeListQuery,
    listCount: attributeListCountQuery,
    findById: attributeFindByIdQuery,
    create: attributeCreateQuery,
    update: attributeUpdateQuery,
    delete: attributeDeleteQuery,
  },
  attributeValue: {
    upsert: { run: attributeValueUpsertQuery },
    deleteMany: attributeValueDeleteManyQuery,
    idList: attributeValueIdListQuery,
    create: attributeValueCreateQuery,
    update: attributeValueUpdateQuery,
    delete: attributeValueDeleteQuery,
    list: attributeValueListQuery,
  },
  // option: {
  //   update: optionUpdateQuery,
  //   create: optionCreateQuery,
  //   delete: optionDeleteQuery,
  //   findById: optionFindByIdQuery,
  // },
  productAttributeValue: {
    list: productAttributeValueListQuery,
    create: productAttributeVAlueInsertQuery,
    delete: productAttributeValueDeleteQuery,
  },
  product: {
    update: productUpdateQuery,
    delete: productDeleteQuery,
    findOne: productFindOneQuery,
    create: productCreateQuery,
    listCount: productListCountQuery,
    list: productListQuery,
    findById: productFindByIdQuery,
  },
  price: {
    upsert: priceUpsertQuery,
  },
};
