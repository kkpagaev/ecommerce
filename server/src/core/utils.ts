import { t } from "./core/trpc";
import { AnyTRPCRouter } from "@trpc/server";
import { ZodSchema } from "zod";

export function withValidation<T extends AnyTRPCRouter>(router: T) {
  const methods = Object.keys(router).filter(
    (key) => !["_def", "createCaller"].includes(key),
  );

  const validator = t.router(
    methods.reduce((acc, key) => {
      const method = router[key];
      const inputs = method._def.inputs;

      const resolver = (req: any) => {
        const body = req.ctx.req.body;
        console.log(body);

        let json = JSON.parse(body);
        json = 0 in json ? json[0] : json;

        const res = inputs.map((i: ZodSchema) => i.safeParse(json));
        return res;
      };
      resolver._def = method._def;

      acc[key] = resolver;
      return acc;
    }, {} as any),
  );

  return t.mergeRouters(
    t.router({
      validator: validator as T,
    }),
    router,
  );
}
