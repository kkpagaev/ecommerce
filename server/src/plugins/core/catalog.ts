import { FastifyInstance } from "fastify";
import fp from "fastify-plugin";
import { Pool } from "pg";

type Categories = ReturnType<typeof Categories>;
function Categories(f: { pool: Pool }) {
  return {
    listCategories: listCategories.bind(null, f.pool),
  };
}

async function listCategories(pool: Pool, limit: number) {
  const res = await pool.query("SELECT * FROM categories LIMIT $1", [limit]);

  return res;
}

type Catalog = {
  categories: Categories;
};

declare module "fastify" {
  export interface FastifyInstance {
    catalog: Catalog;
  }
}

export default fp(async function (f: FastifyInstance) {
  const categories = Categories({ pool: f.pool });

  f.decorate("catalog", {
    categories: categories,
  });
}, {
  dependencies: ["pool", "zod"],
});
