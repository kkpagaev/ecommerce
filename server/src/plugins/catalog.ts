import { FastifyInstance } from "fastify";
import fp from "fastify-plugin";
import { Categories } from "../core/catalog/category";

type Catalog = {
  categories: Categories;
};

declare module "fastify" {
  export interface FastifyInstance {
    catalog: Catalog;
  }
}

export default fp(async function (f: FastifyInstance) {
  const categories = Categories({ pool: f.pool });

  f.decorate("catalog", {
    categories: categories,
  });
}, {
  dependencies: ["pool", "zod"],
});
