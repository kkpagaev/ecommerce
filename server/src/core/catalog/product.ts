import { Translation } from "./i18n";
import { Pool } from "pg";
import { catalogQueries as q } from "./queries";
import slugify from "slugify";
import { tx } from "../../plugins/pool";

export type Products = ReturnType<typeof Products>;
export function Products(f: { pool: Pool }) {
  return {
    createProduct: createProduct.bind(null, f.pool),
    updateProduct: updateProduct.bind(null, f.pool),
  };
}

type CreateProductProps = {
  categoryId: number;
  name: Translation;
  price: number;
  attributes?: Array<number>;
  description?: Translation;
};
async function createProduct(
  pool: Pool,
  input: CreateProductProps,
) {
  const slug = slugify(input.name.uk);

  return tx(pool, async (client) => {
    const product = await q.product.create.run({
      description: input.description,
      name: input.name,
      slug: slug,
      categoryId: input.categoryId,
    }, client).then((res) => res[0]);

    await q.price.upsert.run({
      values: [
        {
          price: input.price,
          type: "default",
          product_id: product.id,
        },
      ],
    }, client);

    await q.productAttributeValue.create.run({
      values: input.attributes?.map((id) => ({
        product_id: product.id,
        attribute_value_id: id,
      })) ?? [],
    }, client);

    return product;
  });
}

type UpdateProductProps = Partial<CreateProductProps>;
async function updateProduct(
  pool: Pool,
  id: number,
  input: UpdateProductProps,
) {
  return null;
}
