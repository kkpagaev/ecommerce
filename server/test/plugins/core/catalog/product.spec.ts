import { describe, expect, it } from "vitest";
import { testDB } from "../../../utils";
import { Products } from "../../../../src/core/catalog/product";
import { catalogQueries as q } from "../../../../src/core/catalog/queries";
import { PoolClient } from "pg";

async function createProducts() {
  const db = await testDB();
  const products = Products({ pool: db.pool });

  return {
    products,
    db,
    [Symbol.asyncDispose]: db[Symbol.asyncDispose],
  };
}

async function fixture(client: PoolClient) {
  const catId = await q.category.create.run({
    name: { uk: "test" },
    slug: "test",
  }, client).then((res) => res[0].id);
  const attributes = await q.attribute.create.run({
    values: [{
      name: { uk: "test" },
      description: { uk: "test" },
    },
    {
      name: { uk: "test2" },
      description: { uk: "test2" },
    }],
  }, client).then((res) => res.map((r) => r.id));
  const attributeValues = await q.attributeValue.create.run({
    values: attributes.map((id) => {
      return {
        value: { uk: "value" + id },
        attributeId: id,
      };
    }),
  }, client).then((res) => res.map((r) => r.id));

  return { catId, attributes, attributeValues };
}

describe("Products", () => {
  describe("createProduct", () => {
    it("should create product with attributes", async () => {
      await using p = await createProducts();
      const fix = await fixture(p.db.client);
      const product = await p.products.createProduct({
        attributes: fix.attributeValues,
        categoryId: fix.catId,
        price: 100,
        name: { uk: "test", en: "test", ru: "test" },
      });

      expect(product).toMatchSnapshot({
        id: expect.any(Number),
      });
    });
  });
});
