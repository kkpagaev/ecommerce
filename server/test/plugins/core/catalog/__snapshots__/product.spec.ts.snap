// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products > createProduct > should create product with attributes 1`] = `
{
  "id": Any<Number>,
}
`;
